distributed-job-manager/src/main/kotlin/ru/fix/distributed/job/manager/Rebalancer.kt

```diff
@@ -0,0 +1,238 @@
+package ru.fix.distributed.job.manager
+
+import org.apache.curator.framework.CuratorFramework
+import org.apache.curator.framework.imps.CuratorFrameworkState
+import org.apache.logging.log4j.kotlin.Logging
+import org.apache.zookeeper.KeeperException.NoNodeException
+import ru.fix.distributed.job.manager.model.AssignmentState
+import ru.fix.distributed.job.manager.model.JobId
+import ru.fix.distributed.job.manager.model.WorkItem
+import ru.fix.distributed.job.manager.model.WorkerId
+import ru.fix.distributed.job.manager.strategy.AssignmentStrategy
+import ru.fix.dynamic.property.api.DynamicProperty
+import ru.fix.zookeeper.transactional.ZkTransaction
+import ru.fix.zookeeper.utils.ZkTreePrinter
+import java.util.*
+
+private const val ASSIGNMENT_COMMIT_RETRIES_COUNT = 3
+
+/**
+ *  If [managerState] allows, performs rebalance according to ZK tree state, using given [assignmentStrategy]
+ * */
+internal class Rebalancer(
+        private val paths: ZkPathsManager,
+        private val curatorFramework: CuratorFramework,
+        private val managerState: DynamicProperty<Manager.State>,
```

Comment:
Why Rebalancer should know about manager state? It provides reassignAndBalanceTasks() method to the Manager. 
If Manager asks - Rebalancer do his job. It is Manager responsibility to call Rebalancer or not to Call. 
Rebalancer do not obligated to know about Manager state or any of Manager internals.